File: skills/fastmail/scripts/caldav-client.ts
import { createDAVClient, DAVCalendar, DAVCalendarObject } from 'tsdav';

const CALDAV_SERVER = 'https://caldav.fastmail.com';

// Type for the client returned by createDAVClient
type CalDAVClientType = ReturnType<typeof createDAVClient> extends Promise<infer T> ? T : never;

export class CalDAVClient {
  private client: CalDAVClientType | null = null;
  private calendars: DAVCalendar[] | null = null;
  
  constructor(
    private username: string,
    private password: string
  ) {}
  
  async connect(): Promise<void> {
    if (this.client) return;
    
    this.client = await createDAVClient({
      serverUrl: CALDAV_SERVER,
      credentials: {
        username: this.username,
        password: this.password,
      },
      authMethod: 'Basic',
      defaultAccountType: 'caldav',
    });
    
    console.error('✅ Connected to Fastmail CalDAV');
  }
  
  async getCalendars(): Promise<DAVCalendar[]> {
    await this.connect();
    
    if (!this.calendars) {
      this.calendars = await this.client!.fetchCalendars();
    }
    
    return this.calendars;
  }
  
  async getCalendarById(calendarId: string): Promise<DAVCalendar | undefined> {
    const calendars = await this.getCalendars();
    return calendars.find(cal => cal.url === calendarId || cal.displayName === calendarId);
  }
  
  async fetchCalendarObjects(
    calendar: DAVCalendar,
    timeRange?: { start: string; end: string }
  ): Promise<DAVCalendarObject[]> {
    await this.connect();
    
    return await this.client!.fetchCalendarObjects({
      calendar,
      timeRange: timeRange || {
        start: new Date().toISOString(),
        end: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000).toISOString(),
      },
    });
  }
  
  async createEvent(
    calendar: DAVCalendar,
    icsData: string,
    eventId: string
  ): Promise<void> {
    await this.connect();
    
    const eventUrl = `${calendar.url}${eventId}.ics`;
    
    await this.client!.createCalendarObject({
      calendar,
      filename: `${eventId}.ics`,
      iCalString: icsData,
    });
  }
  
  async updateEvent(
    calendar: DAVCalendar,
    event: DAVCalendarObject,
    icsData: string
  ): Promise<void> {
    await this.connect();
    
    await this.client!.updateCalendarObject({
      calendarObject: {
        ...event,
        data: icsData,
      },
    });
  }
  
  async deleteEvent(event: DAVCalendarObject): Promise<void> {
    await this.connect();
    
    await this.client!.deleteCalendarObject({
      calendarObject: event,
    });
  }
  
  getClient(): CalDAVClientType {
    if (!this.client) {
      throw new Error('CalDAV client not connected. Call connect() first.');
    }
    return this.client;
  }
}

File: skills/fastmail/scripts/caldav_client.py
"""CalDAV client for Fastmail calendar operations using the caldav library."""

from __future__ import annotations

import sys
from datetime import datetime

import caldav


CALDAV_SERVER = "https://caldav.fastmail.com"


class CalDAVClient:
    def __init__(self, username: str, password: str):
        self._username = username
        self._password = password
        self._client: caldav.DAVClient | None = None
        self._principal: caldav.Principal | None = None
        self._calendars: list[caldav.Calendar] | None = None

    def connect(self) -> None:
        if self._client:
            return

        self._client = caldav.DAVClient(
            url=CALDAV_SERVER,
            username=self._username,
            password=self._password,
        )
        self._principal = self._client.principal()
        print("Connected to Fastmail CalDAV", file=sys.stderr)

    def get_calendars(self) -> list[caldav.Calendar]:
        self.connect()
        if self._calendars is None:
            self._calendars = self._principal.calendars()
        return self._calendars

    def get_calendar_by_id(self, calendar_id: str) -> caldav.Calendar | None:
        calendars = self.get_calendars()
        for cal in calendars:
            if str(cal.url) == calendar_id or cal.name == calendar_id:
                return cal
        return None

    def fetch_calendar_objects(
        self,
        calendar: caldav.Calendar,
        time_range: dict[str, str] | None = None,
    ) -> list:
        self.connect()
        if time_range:
            start = datetime.fromisoformat(time_range["start"])
            end = datetime.fromisoformat(time_range["end"])
            return calendar.date_search(start=start, end=end, expand=True)
        return calendar.events()

    def create_event(self, calendar: caldav.Calendar, ics_data: str, _event_id: str = "") -> None:
        self.connect()
        calendar.save_event(ics_data)

    def update_event(self, calendar: caldav.Calendar, event: object, ics_data: str) -> None:
        self.connect()
        event.data = ics_data
        event.save()

    def delete_event(self, event: object) -> None:
        self.connect()
        event.delete()
